### PR TITLE
Graph break on amd sdpa

### PR DESCRIPTION
**Pull request by OrenLeung**
- [x] graph break on sdpa 
- [x] todo, graph break sdpa for only amd
- [x] simplified MFU calucation
- [x] simpify profiling loop
- [x] fused AdamW
- [x] gradient_as_bucket_view for DDP to lower memory count
- [x] add compile mode as cli arg

### CODE REVIEW

**Review by kimbochen — APPROVED** [review]
Really great, thank you for the update.

A few questions:
- The sdpa wrapper in `gpt.py` is different from the one in `llama.py`, is this intentional?
- Does measuring FLOP/s while profiling cause overhead (`torch.cuda.synchronize`) significant enough to affect the profiling results?

**Comment by OrenLeung** [comment]
> Really great, thank you for the update.
> 
> 
> 
> A few questions:
> 
> - The sdpa wrapper in `gpt.py` is different from the one in `llama.py`, is this intentional?

Good catch. Will fix in a bit.

> - Does measuring FLOP/s while profiling cause overhead (`torch.cuda.synchronize`) significant enough to affect the profiling results?

I don't believe it does affect profiling results to do CUDA events and syncs when profiling. Having sync (loss.item()) at the end of a training step is pretty realistic anywayd
